File: Proj4a/Questão 4/Insertion-sort/insertion-sort.java
import java.util.Random;

public class InsertionSortPerformance {

    void insertionSort(int arr[]) {
        int n = arr.length;
        for (int i = 1; i < n; ++i) {
            int key = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j = j - 1;
            }
            arr[j + 1] = key;
        }
    }

    void printArray(int arr[]) {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }

    long measureExecutionTime(int[] arr) {
        long startTime = System.nanoTime();
        insertionSort(arr);
        long endTime = System.nanoTime();
        return endTime - startTime; // Tempo em nanossegundos
    }

    void generateAndTestArrays(int maxSize) {
        Random random = new Random();
        System.out.println("Tamanho do array, Tempo médio (em nanossegundos)");

        for (int size = 10; size <= maxSize; size *= 10) {
            long totalTime = 0;

            // Executar várias vezes e tirar a média
            int[] arr = new int[size];
            int numTrials = Math.max(10000000 / size, 1); // ajuste para o tamanho do array

            for (int trial = 0; trial < numTrials; trial++) {
                // Preencher o array com números aleatórios
                for (int i = 0; i < size; i++) {
                    arr[i] = random.nextInt();
                }

                totalTime += measureExecutionTime(arr);
            }

            long averageTime = totalTime / numTrials;
            System.out.println(size + ", " + averageTime);
        }
    }

    public static void main(String args[]) {
        InsertionSortPerformance ob = new InsertionSortPerformance();
        ob.generateAndTestArrays(1000000); // Tamanho máximo ajustado conforme necessário
    }
}

File: Proj4a/Questão 4/Insertion-sort/insertion-sort.pl
% Predicado para gerar uma lista de tamanho N com números aleatórios
generate_list(N, List) :-
    length(List, N),
    maplist(random_between(1, 1000), List). % números aleatórios entre 1 e 1000

% Predicado para gerar listas para diferentes tamanhos até um máximo especificado
generate_lists(MaxSize, Lists) :-
    findall(List, (between(1, MaxSize, N), generate_list(N, List)), Lists).

% insertion sort implementation
insert(X, [], [X]).
insert(X, [Y|T], [X,Y|T]) :- X =< Y.
insert(X, [Y|T], [Y|NT]) :- X > Y, insert(X, T, NT).

insertion_sort([], []).
insertion_sort([H|T], Sorted) :-
    insertion_sort(T, ST),
    insert(H, ST, Sorted).

% Predicado para medir o tempo de execução médio do insertion_sort para uma lista
average_time_insertion_sort(List, AvgTime) :-
    statistics(runtime, _),
    insertion_sort(List, _Sorted),
    statistics(runtime, [_, Time]),
    length(List, Len),
    AvgTime is Time / Len.

% Predicado para medir o tempo de execução médio do sort embutido para uma lista
average_time_builtin_sort(List, AvgTime) :-
    statistics(runtime, _),
    sort(List, _Sorted),
    statistics(runtime, [_, Time]),
    length(List, Len),
    AvgTime is Time / Len.

% Predicado para medir o tempo médio de execução do insertion_sort para várias listas
average_times_insertion_sort(Lists, AvgTimes) :-
    maplist(average_time_insertion_sort, Lists, AvgTimes).

% Predicado para medir o tempo médio de execução do sort embutido para várias listas
average_times_builtin_sort(Lists, AvgTimes) :-
    maplist(average_time_builtin_sort, Lists, AvgTimes).

% Predicado para realizar a comparação de desempenho
compare_performance(MaxSize) :-
    generate_lists(MaxSize, Lists),
    average_times_insertion_sort(Lists, AvgTimesInsertion),
    average_times_builtin_sort(Lists, AvgTimesBuiltIn),
    write('Tamanho da lista, Insertion Sort (ms/elemento), Sort embutido (ms/elemento)'), nl,
    write_separator(80),
    print_results(Lists, AvgTimesInsertion, AvgTimesBuiltIn, 1).

% Predicado auxiliar para imprimir resultados formatados
print_results([], [], [], _).
print_results([List|RestLists], [TimeInsertion|RestInsertion], [TimeBuiltIn|RestBuiltIn], Index) :-
    length(List, Len),
    format('| ~|~`0t~d~2+ | ~|~`0t~2f~6+ | ~|~`0t~2f~6+ |~n', [Len, TimeInsertion, TimeBuiltIn]),
    NextIndex is Index + 1,
    print_results(RestLists, RestInsertion, RestBuiltIn, NextIndex).

% Predicado auxiliar para imprimir separador
write_separator(0) :- !, nl.
write_separator(N) :-
    N > 0,
    write('-'),
    N1 is N - 1,
    write_separator(N1).

% Exemplo de uso:
% ?- compare_performance(1000).

% Para visualizar os tempos médios obtidos, ajuste o limite de tamanho de lista conforme necessário.
